src/libfetchers/git-utils.cc

```diff
@@ -964,11 +985,27 @@ struct GitFileSystemObjectSinkImpl : GitFileSystemObjectSink
             git_tree_entry_filemode(entry));
     }
 
-    Hash sync() override {
+    Hash sync() override
+    {
         updateBuilders({});
 
         auto [oid, _name] = popBuilder();
 
+        /* If the root directory contains a single entry that is a
+           directory or a non-executable regular file, return that as
+           the top-level object. We don't do this for executables
+           because they don't have a tree hash in the Git object
+           model. */
+        auto _tree = lookupObject(*repo, oid, GIT_OBJECT_TREE);
+        auto tree = (const git_tree *) &*_tree;
+
+        if (git_tree_entrycount(tree) == 1) {
+            auto entry = git_tree_entry_byindex(tree, 0);
+            auto mode = git_tree_entry_filemode(entry);
+            if (mode == GIT_FILEMODE_BLOB || mode == GIT_FILEMODE_TREE)
+                oid = *git_tree_entry_id(entry);
+        }
+
```

Comment:
This isn't a git responsibility.
Shouldn't this be done in the tarball fetcher? It's encoding tarball-specific logic.

See also https://github.com/NixOS/nix/pull/11195/files#diff-7b407c6754b02bc382cb4344e00c566919584fd7d75ab2624de5d4f6b32bab46R59 where I've noticed this during testing. That test needs to be broken, and can be fixed to test the right thing by removing all of its `foo-1.1/` occurrences.

Reply:
Done, I've split this into a separate method.